src/source-updater.js

```diff
@@ -162,67 +371,121 @@ export default class SourceUpdater {
    * @return {Boolean} the updating status of the SourceBuffer
    */
   updating() {
-    // we are updating if the sourcebuffer is updating or
-    return !this.sourceBuffer_ || this.sourceBuffer_.updating ||
-      // if we have a pending callback that is not our internal noop
-      (!!this.pendingCallback_ && this.pendingCallback_ !== noop);
+    // the audio source buffer is updating
+    if (updating('audio', this)) {
+      return true;
+    }
+
+    // the video source buffer is updating
+    if (updating('video', this)) {
+      return true;
+    }
+
+    return false;
   }
 
   /**
-   * Set/get the timestampoffset on the SourceBuffer
+   * Set/get the timestampoffset on the audio SourceBuffer
    *
    * @return {Number} the timestamp offset
    */
-  timestampOffset(offset) {
-    if (typeof offset !== 'undefined') {
-      this.queueCallback_(() => {
-        this.sourceBuffer_.timestampOffset = offset;
-        this.runCallback_();
+  audioTimestampOffset(offset) {
+    if (typeof offset !== 'undefined' &&
+        this.audioBuffer &&
+        // no point in updating if it's the same
+        this.audioTimestampOffset_ !== offset) {
+      pushQueue({
+        type: 'audio',
+        sourceUpdater: this,
+        action: actions.timestampOffset(offset),
+        name: 'timestampOffset'
       });
-      this.timestampOffset_ = offset;
+      this.audioTimestampOffset_ = offset;
     }
-    return this.timestampOffset_;
+    return this.audioTimestampOffset_;
   }
 
   /**
-   * Queue a callback to run
+   * Set/get the timestampoffset on the video SourceBuffer
+   *
+   * @return {Number} the timestamp offset
    */
-  queueCallback_(callback, done) {
-    this.callbacks_.push([callback.bind(this), done]);
-    this.runCallback_();
+  videoTimestampOffset(offset) {
+    if (typeof offset !== 'undefined' &&
+        this.videoBuffer &&
+        // no point in updating if it's the same
+        this.videoTimestampOffset !== offset) {
+      pushQueue({
+        type: 'video',
+        sourceUpdater: this,
+        action: actions.timestampOffset(offset),
+        name: 'timestampOffset'
+      });
+      this.videoTimestampOffset_ = offset;
+    }
+    return this.videoTimestampOffset_;
   }
 
-  /**
-   * Run a queued callback
-   */
-  runCallback_() {
-    let callbacks;
+  audioQueueCallback(callback) {
+    if (this.audioBuffer) {
+      pushQueue({
+        type: 'audio',
+        sourceUpdater: this,
+        action: actions.callback(callback),
+        name: 'callback'
+      });
+    }
+  }
 
-    if (!this.updating() &&
-        this.callbacks_.length &&
-        this.started_) {
-      callbacks = this.callbacks_.shift();
-      this.pendingCallback_ = callbacks[1];
-      callbacks[0]();
+  videoQueueCallback(callback) {
+    if (this.videoBuffer) {
+      pushQueue({
+        type: 'video',
+        sourceUpdater: this,
+        action: actions.callback(callback),
+        name: 'callback'
+      });
     }
   }
```

Comment:
I think this file will eventually need some updates. Some places we have function that pass a `type` and are generic for audio and video. Then others functions are prepended with `audio` or `video` (IE `audioTimestampOffset` vs `videoTimestampOffset`) and they do nearly the same thing.